Draft: Add back skip buttons and more custom actions on Android Auto

awbooze opened the pull request:
Closes #5481.

Remove ACTION_SKIP_TO_NEXT from showing because it's hard-coded to appear on the near right for Android Auto. Replace it with a custom change playback speed action and skip to next actions in order to fill out the actions list. I haven't wired up the playback speed action because I wanted to make sure that was the direction we wanted to go for that last action slot. It's the fifth action in the widget, so I assumed it was a leading candidate, though.

This PR has some potential downsides. It removes the "Skip to next" action from Wear OS - that could be a future area of improvement (although, for example, Google Podcasts doesn't add any actions on Wear OS). We'd also have to implement a new way of switching between playback speeds because we can't have people looking at the screen to pick a playback speed. However, when it's time to target Android 13, it should be instantly compatible with [the new Android media controls](https://developer.android.com/about/versions/13/behavior-changes-13#playback-controls).

Screenshots:
![AntennaPod Fix Full Screen](https://user-images.githubusercontent.com/42682253/187114297-314374d9-8e26-4ae2-8baa-ea8428cd95a9.png)
![AntennaPod Fix Minimized to bottom of Android Auto Taskbar](https://user-images.githubusercontent.com/42682253/187114313-a83a3b3a-9108-4470-951f-5518884642d2.png)
![AntennaPod Fix API 33 Media Notification](https://user-images.githubusercontent.com/42682253/187114394-f3a87187-5044-4d20-96ae-9114be869d12.png)

keunes commented:
> This PR has some potential downsides. It removes the "Skip to next" action from Wear OS - that could be a future area of improvement

I would be careful with that. We'd effectively be removing existing functionality that seems to be used by people (as mentioned in #3035).

awbooze commented:
I downloaded a couple of podcast apps and it looked like they had a way to do it, so I did some more research and eventually figured out that WearMediaSession wasn't just a no-op. I changed it so it could make the fast forward and rewind actions appear on Wear OS as well (see screenshot), although only for the Play flavor. It looks like we're only using `android.support.wearable.media` as a dependency for the Google Play version - is there a reason we couldn't use it for both versions?
![A screenshot of Wear OS with all three types of Media Controls showing](https://user-images.githubusercontent.com/42682253/190936348-9065def4-4bcf-4b2f-9028-cae39940c87b.png)

I also had some weird problems with icons changing or disappearing on Android Auto when flipping between playDebug and freeDebug builds - is that to be expected?

awbooze commented:
I double-checked both builds in my car today and both worked as expected. I guess it was the desktop head unit being weird?

awbooze commented:
Also related: PR #5972.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1878,6 +1906,21 @@ public void onCustomAction(String action, Bundle extra) {
+                int speedPosition = selectedSpeeds.indexOf(mediaPlayer.getPlaybackSpeed());
+
+                float newSpeed;
+                if (speedPosition == selectedSpeeds.size() - 1) {
+                    // This is the last element. Wrap instead of going over the size of the list.
+                    newSpeed = selectedSpeeds.get(0);
```
This will crash if there are no speed presets saved.

awbooze replied:
Yep, that will - I'll fix it.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1854,6 +1876,12 @@ public void onSeekTo(long pos) {
         @Override
         public void onSetPlaybackSpeed(float speed) {
             Log.d(TAG, "onSetPlaybackSpeed()");
+            PlaybackPreferences.setCurrentlyPlayingTemporaryPlaybackSpeed(speed);
+            if (currentMediaType == MediaType.VIDEO) {
+                UserPreferences.setVideoPlaybackSpeed(speed);
+            } else {
+                UserPreferences.setPlaybackSpeed(speed);
+            }
```
This copied from PlaybackController, right? Would it make sense to move the code to `setSpeed`? Having multiple copies of the same code will probably cause issues later, otherwise.

awbooze replied:
Yeah, that would make sense. I think my worry is that I'm still not sure where SpeedChangedEvents from the PlaybackController are handled, so that might mess something up down the line.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1837,6 +1858,7 @@ public void onSkipToNext() {
             UiModeManager uiModeManager = (UiModeManager) getApplicationContext()
                     .getSystemService(Context.UI_MODE_SERVICE);
             if (UserPreferences.getHardwareForwardButton() == KeyEvent.KEYCODE_MEDIA_NEXT
+                    || Build.VERSION.SDK_INT >= Build.VERSION_CODES.S
```
Doesn't this make the "Forward button skips" setting unusable?

awbooze replied:
Yes, I think it would. Looking at the code now, I think it would be better to just call mediaPlayer.skip() when the custom skip action is called, since we always want to skip in that case.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1225,33 +1228,51 @@ private void updateMediaSession(final PlayerStatus playerStatus) {
-                        android.R.drawable.ic_media_ff);
-                WearMediaSession.mediaSessionSetExtraForWear(mediaSession);
-            }
+            // Prevent next button from showing up unless we aren't showing the custom skip action
```
So the custom skip action is needed because of the order of the buttons?

awbooze replied:
Yes. In my testing, there was no way to move the default skip action from its default position just to the right of the play button to any other position without removing it and creating a custom skip action instead.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1225,33 +1228,51 @@ private void updateMediaSession(final PlayerStatus playerStatus) {
-        if (uiModeManager.getCurrentModeType() == Configuration.UI_MODE_TYPE_CAR) {
+        UiModeManager uiModeManager = (UiModeManager) getApplicationContext().getSystemService(Context.UI_MODE_SERVICE);
+        int currentModeType = uiModeManager.getCurrentModeType();
+        Log.d(TAG, "UI Mode CurrentModeType = " + currentModeType);
```
I'm not sure this needs to be logged in the production code

ByteHamster reviewed (commented):
So sorry for the long delay! The email about this PR somehow moved towards the bottom of my inbox. Now that I have a bit more time to go through some of the older AntennaPod emails again, I noticed that your PR is still on my to-do list.

I just tried it and had a quick look at the code. I also fixed the merge conflict with `develop`. Unfortunately, on my phone, it does not look like in your screenshots (even before the merge). I don't get a playback speed button, and the skip buttons are both to the right side of the progress bar. Any idea what could be going on there?

> I went ahead and implemented rotating through playback speeds in order. However, when it's time to target Android 13, it should be instantly compatible with [the new Android media controls](https://developer.android.com/about/versions/13/behavior-changes-13#playback-controls).

Nice! In the meantime, we now do target Android 13 (API 31) :) For some reason, I still don't see the large play button, though. Is there any additional change needed?

> It looks like we're only using android.support.wearable.media as a dependency for the Google Play version - is there a reason we couldn't use it for both versions?

My assumption would be that the library is closed-source and we can therefore not publish it on F-Droid. Though, I'm not 100% sure.

> I also had some weird problems with icons changing or disappearing on Android Auto when flipping between playDebug and freeDebug builds - is that to be expected?

Hmm, no, not really...

awbooze commented:
Thanks for getting back to this! I won't be able to get to everything tonight, but I can hopefully work on this over the next week or so.

> I just tried it and had a quick look at the code. I also fixed the merge conflict with `develop`. Unfortunately, on my phone, it does not look like in your screenshots (even before the merge). I don't get a playback speed button, and the skip buttons are both to the right side of the progress bar. Any idea what could be going on there?

That's really odd. I'm still getting UI that matches the screenshots on the [desktop head unit emulator](https://developer.android.com/training/cars/testing). I'm not sure what would cause that, since Android Auto shouldn't change based on the phone.

> > I went ahead and implemented rotating through playback speeds in order. However, when it's time to target Android 13, it should be instantly compatible with [the new Android media controls](https://developer.android.com/about/versions/13/behavior-changes-13#playback-controls).
> 
> Nice! In the meantime, we now do target Android 13 (API 31) :) For some reason, I still don't see the large play button, though. Is there any additional change needed?

API 31 is Android 12; we'd need to target API 33 to get the new player interface on Android 13.

> > It looks like we're only using android.support.wearable.media as a dependency for the Google Play version - is there a reason we couldn't use it for both versions?
> 
> My assumption would be that the library is closed-source and we can therefore not publish it on F-Droid. Though, I'm not 100% sure.

In the mean time, I now have a Wear OS 3 watch, so I can confirm that android.support.wearable.media doesn't work on Wear OS 3 (although it works perfectly on Wear OS 2). I'll do a bit more research into that and see if there's a solution that works on modern devices.

TacoTheDank commented:
>In the mean time, I now have a Wear OS 3 watch, so I can confirm that android.support.wearable.media doesn't work on Wear OS 3 (although it works perfectly on Wear OS 2). I'll do a bit more research into that and see if there's a solution that works on modern devices.

Some more details if you want: https://github.com/AntennaPod/AntennaPod/pull/5972#issuecomment-1190803127

ByteHamster commented:
>> Nice! In the meantime, we now do target Android 13 (API 31) :) For some reason, I still don't see the large play button, though. Is there any additional change needed?
>
> API 31 is Android 12; we'd need to target API 33 to get the new player interface on Android 13.

Ah, right. Sorry. When I compile for API 33, I do get the large play button and also the playback speed button.

Kind of weird that it doesn't make full use of the MediaSession before targeting 33. I mean, didn't older Android versions already use the media session? Now when targeting 31, it only uses the actions from the notification and ignores the media session. Anyway. I tried to add the playback speed button to the notification as well, but I think that might be too much. Then the notification gets really crowded (see screenshot below). So I guess the current behavior of only having the speed button on Android 13 (when the play button is positioned differently) is actually fine.

<img width="250" src="https://user-images.githubusercontent.com/5811634/218252055-24b7462f-335b-47b0-9373-5984b102c8bc.jpg" />

So I guess this PR only misses the few comments I left above, then it can be merged :)

> I'll do a bit more research into that and see if there's a solution that works on modern devices.

That would be awesome! I don't own a wearable device, so I don't have a high intrinsic motivation to get it fixed (and my AntennaPod inbox is full enough anyway). So pull requests would be very welcome.

awbooze commented:
Current Status:
- Issues identified have been corrected.
- This PR is *better* for:
    - Users using Android Auto
    - Users using Android 13 (once we target it)
    - Users using Wear OS 2
- This PR is *worse* for:
    - Users using Wear OS 3 (removes "next" button support)

It looks like it should be possible to use [AndroidX Media3](https://developer.android.com/jetpack/androidx/releases/media3) to support Wear OS 3, as well as Android Auto and notifications (see https://github.com/androidx/media/issues/38 and https://github.com/androidx/media/issues/216). However, Media3 is still in beta, has issues, and looks like it would require us to jump multiple dependency versions. I would rather merge this and start on a new PR than bundle those major changes into one PR. If that makes sense, I would not close #3035 with this PR and I could go ahead and create a draft for those changes when I have time.

ByteHamster reviewed (approved):
Sounds good. Ready for merging @awbooze?

awbooze commented on core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java:
```diff
@@ -1159,42 +1162,60 @@ private void updateMediaSession(final PlayerStatus playerStatus) {
+
+        // Workaround bug in Android Auto custom actions described at https://issuetracker.google.com/issues/207389461 by checking for API 31+
+        // as well.
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S || currentModeType == Configuration.UI_MODE_TYPE_CAR) {
```
I checked on this today. On Android 10 and below with Wear OS 2, the screen will look like this, which is better than the previous code for users using Wear OS 2. I'm satisfied with this result and ready to merge.

![A screenshot of a Wear OS emulator displaying the playback interface](https://user-images.githubusercontent.com/42682253/220241525-a34b10ab-0ce2-4ee1-acde-3c204f92892b.png)

ByteHamster replied:
Hmm, this screenshot doesn't look very optimal to me. What did it show before? Fast-Forward is probably a lot more important than skip.

awbooze replied:
With the current build of AntennaPod, it shows the skip button without the rewind button with all versions of Android on Wear OS 2. We could move to always show the rewind and fast forward buttons (see my comment from September 18) instead by removing the if statement altogether and never using the built-in skip button. That would be easy to change if you think that's better.

ByteHamster replied:
> We could move to always show the rewind and fast forward buttons (see my comment from September 18) instead by removing the if statement altogether and never using the built-in skip button.

That would be great!

awbooze replied:
Done.

awbooze commented:
Yes, I'm ready to merge @ByteHamster.

ByteHamster commented:
Thanks! Let's hope that this will not break anything on some other platforms. Dealing with the media sessions is always a bit of a pain...

awbooze commented:
Hopefully not, but if you find anything, feel free to tag me on the ticket or forums and I can take a look!

ByteHamster commented:
This change is not included in AntennaPod 3.0.x yet. It will be released in 3.1.x.

awbooze commented:
Thanks @ByteHamster for the update. Do you know approximately when 3.1.x is targeted for release?

ByteHamster commented:
No, I don't know. 3.0.0 isn't even released fully yet, and there still is a huge rewrite of the download logic pending before 3.1.0 can be anywhere near releasing

awbooze commented:
Okay, that's fine. Just wondering if you knew.

tony1661 commented:
Hi, I would love this change. Any news on the release of 3.1? 🙂 

Thanks for the hard work on fixing this!

keunes commented:
@tony1661 Due to summer everything is in a bit of a slow down these months. Version 3.1 is currently in beta, and will probably be released some point between late August and mid September.
